EPPs/common.py

```diff
@@ -173,6 +183,105 @@ def patch_entry(cls, *args, **kwargs):
         return cls._rest_interaction(rest_communication.patch_entry, *args, **kwargs)
 
 
+class GenerateHamiltonInputEPP(StepEPP):
+
+    # define the rows and columns in the input plate (standard 96 well plate pattern)
+    plate_rows = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']
+    plate_columns = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']
+    csv_column_headers = None
+    output_file_name = None
+    permitted_input_containers = None
+    permitted_output_containers = None
+
+    def __init__(self, argv=None):
+        """ additional argument required for the location of the Hamilton input file so def __init__ customised."""
+        super().__init__(argv)
+        self.hamilton_input = self.cmd_args.hamilton_input
+        self.shared_drive_path = os.path.abspath(self.cmd_args.shared_drive_path)
+
+        assert self.csv_column_headers is not None, 'csv_column_headers needs to be set by the child class'
+        assert self.output_file_name is not None, 'output_file_name needs to be set by the child class'
+        assert self.permitted_input_containers is not None, 'number of permitted input containers needs to be set by the child class'
+        assert self.permitted_output_containers is not None, 'number of permitted output containers needs to be set by the child class'
+
+
+    @staticmethod
+    def add_args(argparser):
+        argparser.add_argument(
+            '-i', '--hamilton_input', type=str, required=True, help='Hamilton input file generated by the LIMS'
+        )
+        argparser.add_argument(
+            '-d', '--shared_drive_path', type=str, required=True,
+            help='Shared drive path location for Hamilton input file'
+        )
+
+    @staticmethod
+    def write_csv(filename, csv_array):
+        """Write the list of list to the file provided as a csv file"""
+        with open(filename, 'w') as csvFile:
+            writer = csv.writer(csvFile)
+            writer.writerows(csv_array)
+
+    @cached_property
+    def input_container_names(self):
+        """The name of containers from input artifacts"""
+        containers = []
```

Comment:
Seems a bit roundabout to build a list, then cast it to a set, then back to a list - would be simpler to just build a set and cast it at the end. I also don't see the need for `frozenset` here